src/lib.rs

```diff
@@ -40,24 +48,69 @@ pub struct FastraceService<S> {
     service: S,
 }
 
-impl<S, Body> Service<Request<Body>> for FastraceService<S>
-where S: Service<Request<Body>>
+use axum::extract::Request;
+
+impl<S> Service<Request> for FastraceService<S>
+where
+    S: Service<Request, Response = Response> + Send + 'static,
+    S::Future: Send + 'static,
 {
     type Response = S::Response;
     type Error = S::Error;
-    type Future = fastrace::future::InSpan<S::Future>;
+    type Future =
+        Pin<Box<dyn Future<Output = Result<Self::Response, Self::Error>> + Send + 'static>>;
 
     fn poll_ready(&mut self, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
         self.service.poll_ready(cx)
     }
 
-    fn call(&mut self, req: Request<Body>) -> Self::Future {
+    fn call(&mut self, req: Request) -> Self::Future {
         let headers = req.headers();
         let parent = headers
             .get(TRACEPARENT_HEADER)
             .and_then(|traceparent| SpanContext::decode_w3c_traceparent(traceparent.to_str().ok()?))
             .unwrap_or(SpanContext::random());
-        let root = Span::root(req.uri().to_string(), parent);
-        self.service.call(req).in_span(root)
+
+        let span_name = get_request_span_name(&req);
+        let root = Span::root(span_name, parent);
+
+        root.add_properties(|| {
```

Comment:
Prefer to add_properties multiple times instead of constructing a vec, e.g.,

```rust
root.add_properties(||  [...] );
if let Some(route) = ... {
    root.add_property(||  ... );
}
```